src/simple_pathfinding.cpp

```diff
@@ -0,0 +1,354 @@
+            if( open[n] == 0 || open[n] > priority ) {
+                dirs[n] = ( static_cast<int>( dir ) + 2 ) % 4;
+                if( open[n] != 0 ) {
+                    while( nodes[i].top().pos != p ) {
+                        nodes[1 - i].push( nodes[i].top() );
+                        nodes[i].pop();
+                    }
+                    nodes[i].pop();
+                    if( nodes[i].size() > nodes[1 - i].size() ) {
+                        i = 1 - i;
+                    }
+                    while( !nodes[i].empty() ) {
+                        nodes[1 - i].push( nodes[i].top() );
+                        nodes[i].pop();
+                    }
+                    i = 1 - i;
+                }
+                open[n] = priority;
+                nodes[i].emplace( p, dir, priority );
+            }
+        }
+    }
+    return res;
+}
+
+namespace
+{
+
+const tripoint &direction_to_tripoint( direction dir )
+{
+    switch( dir ) {
+        case direction::EAST:
+            return tripoint_east;
+        case direction::SOUTH:
+            return tripoint_south;
+        case direction::WEST:
+            return tripoint_west;
+        case direction::NORTH:
+            return tripoint_north;
+        case direction::ABOVECENTER:
+            return tripoint_above;
+        case direction::BELOWCENTER:
+            return tripoint_below;
+        default:
+            debugmsg( "Unexpected direction: %d", static_cast<int>( dir ) );
+            return tripoint_zero;
+    }
+}
+
+bool is_horizontal( direction dir )
+{
+    switch( dir ) {
+        case direction::EAST:
+        case direction::SOUTH:
+        case direction::WEST:
+        case direction::NORTH:
+            return true;
+        default:
+            return false;
+    }
+}
+
+// The address of a navigation node, a compressed tripoint relative to the starting
+// point, i.e. the start is always (0, 0, 0).
+// NOLINTNEXTLINE(cata-xy)
+struct node_address {
+    int16_t x;
+    int16_t y;
+    int8_t z;
+    explicit node_address( const tripoint &p ) : x( p.x ), y( p.y ), z( p.z ) {}
+    bool operator== ( const node_address &other ) const {
+        return x == other.x && y == other.y && z == other.z;
+    }
+    tripoint_abs_omt to_tripoint( const tripoint_abs_omt &origin ) const {
+        return origin + tripoint( x, y, z );
+    }
+    node_address operator+ ( const tripoint &p ) const {
+        node_address ret = *this;
+        ret.x += p.x;
+        ret.y += p.y;
+        ret.z += p.z;
+        return ret;
+    }
+    node_address displace( direction dir ) const {
+        return *this + direction_to_tripoint( dir );
+    }
+};
+
+struct node_address_hasher {
+    std::size_t operator()( const node_address &addr ) const {
+        std::size_t val = addr.x;
+        val = ( val << 16 ) + addr.y;
+        val = ( val << 16 ) + static_cast<int>( addr.z );
+        return std::hash<std::size_t> {}( val );
+    }
```

Comment:
I realize this is an old PR but I happened to be looking at this code and noticed that this is a really bad hash function on 32-bit platforms because it doesn't depend on `x`.

Reply:
Yeah, I guess the type on L195 should be changed to std::int64_t

Reply:
That wouldn't help, hash values always get converted to `size_t` eventually.

Reply:
In making that reply I made the incorrect assumption that std::hash invoked on a 64-bit integer on a 32-bit system would do something more intelligent than just take half of the bits and discard the rest. I also noticed that the same mistake is present in vehicle_autodrive.cpp.

I can fix this by scrambling the bits, either by using a bit mixing function or using hash_combine on the 2 halves.
I'd like to use template magic so that the scrambling only happens on 32-bit systems. But the idea of duplicating that code in 2 cpp files doesn't sit well with me, so I guess I need to add something in hash_utils.h.
I'm thinking of something like "std::size_t cata::hash64(std::uint64_t)", with 2 templated specializations based on the size of size_t, one of which is the identity function for the case of size_t being 8 bytes or more.
I'm not very good at template meta-programming so it might take me a while to figure out the incantation.

Reply:
Please take a look at #53019